### PR TITLE
Native touch passthrough, better keyboard toggle, better remote typing, anti-unexpected-session-exit &  touchpoint velocity (sensitivity) tweak

### PR DESCRIPTION
**Pull request by TrueZhuangJia**
- ### Robust native multi-touch pass-through.
  testt5
   <br><br>

- ### Stream view lifting for local keyboard, prevent remote typing view from being blocked by local keyboard:
  I did this because iOS floating keyboard sucks.
  image
  ViewliftExplained - Copy
  Use Gesture to let the software know where the remote input field is, then stream view will be lifted appropriately (not going to be covered by keyboard). Here's a gif example: 
  testt7
   <br><br>

- ###  Refactored tap gesture recognition for reliable local keyboard toggle.
   The official version of Moonlight-iOS triggers the local keyboard switch when it detects three touch points on the screen, a mechanism that makes three-point dragging completely ineffective. <br> I also tried  UITapGestureRecognizer provided by iOS native API: the success rate of gesture recognition would decrease, and even continuous failures could occur. Therefore, I rewrote a TapGestureRecognizer with a recognition rate of almost 100% that does not affect multi-point dragging.
   To prevent local keyboard from being invoked unexpectedly, I add the required number of fingers for tap recognizer to the setting menu. It is recommended to set 3-finger triggering on phones and 4 or more on tablets (as your palm might touch the screen, forming a third touch point when using a tablet).
   image
   Here's a *.gif example of setting "Fingers to Tap" to 4 on my ipad mini6: 
   testt5
<br><br>

- ### Configurable local keyboard toolbar:
  image
   <br><br>

-  ### Touch Pointer Velocity (Sensitivity) Tweak 
    image
   The pointer velocity tweak zone is on the right side of screen with configurable width, while on the rest part of screen, original coordinates of pointer will be passed through to the remote PC. <br> Useful for tweaking view rotation in native touch games.
testt6
   <br><br>
   

- ### Refactored Session Exit Gesture Recognition to Prevent Accidental Exits.
   Due to the Mihoyo game UI's directional wheel on the left being close to the screen edge, the original exit gesture recognition was frequently triggered during intense touch operations. You cannot imagine the despair of suddenly exiting from desktop in the Genshin Impact Spiral Abyss. I had to write a recognizer myself to replace the iOS native API, requiring a swipe from the screen edge to a certain distance to trigger the exit to the desktop. The trigger threshold for the swipe distance and which edge to start the swipe from have been added to the settings menu: 
image
longSwipeToExit
   <br><br>

- ### Improved Mouse Right Click & Wheel Scroll Gesture in Trackpad (Relative Touch) Mode
  In original version, both right click & wheel scroll are triggered by 2 finger gesture, causes competing. Right click could fail and result in unexpected move of scroll view. This is resolved by utilizing my CustomTapGestureRecognizer, and a time threshold mechnism to prevent the 2 gestures from competing each other.

### CODE REVIEW

**Comment by ChrisKimZHT** [comment]
Multi-touch is a fantastic feature for mobile phone game streaming. Looking forward to seeing this PR merged into the main branch.

File: Limelight/Input/StreamView.m
Comment (TrueZhuangJia) [inline Limelight/Input/StreamView.m:1]: Absolute touch mode is restored for Mac OS host.
image
3 touch modes available now.

**Comment by stefanilijev97** [comment]
This is a very good feature and many people are looking forward to this.. but it seems that Moonlight team is sleeping and nothing is done to improve the product further (no updates on iPadOS at least 3 months and still going).

**Comment by TrueZhuangJia** [comment]
Seems that the original moonlight-iOS is not going to be updated, try my modified version for much more improvements
https://github.com/Moonlight-the-Fried-Fish/moonlight-ios-NativeMultiTouchPassthrough

**Comment by xXJSONDeruloXx** [comment]
Why won't they merge??

**Comment by TrueZhuangJia** [comment]
> Why won't they merge??

no idea. but why would they？ they don't have to merge anything.

**Comment by stefanilijev97** [comment]
> > Why won't they merge??
> 
> no idea. but why would they？ they don't have to merge anything.

Because the project is not going anywhere and there are people actively working on solving people’s problems and creating overall better product. Devs that worked originally don’t upload any new things and are just lazy, so somebody like this needs to step in to solve the problem.

**Comment by TrueZhuangJia** [comment]
> > > Why won't they merge??
> > 
> > 
> > no idea. but why would they？ they don't have to merge anything.
> 
> Because the project is not going anywhere and there are people actively working on solving people’s problems and creating overall better product. Devs that worked originally don’t upload any new things and are just lazy, so somebody like this needs to step in to solve the problem.

i don't blame them anyway ~ 'cause i'm probably about to become lazy too.  ⸜( •ᴗ• )⸝

**Pull request closed without merging**